Fix slow start when network shares are unavailable (#1487138)

sblaisot opened the pull request:
Fix slow start when network shares are unavailable (#1487138)

sblaisot committed:
fix slow start when network share are unavailable (#1487138)

daschuer commented:
Thank you for the fix. It will work for me. Please add a comment that drive.canonicalPath() takes long if a network share is not available. 

Please file also a new Launchpad bug that describes the remaining issue for other canonicalPath() calls. 

Does every substantial canonicalPath() call suffer the delay? If yes, a possible solution would be to blacklist unavailable drives. This might be done by a background worker that monitors all dives and updates the blacklist. The original  canonicalPath() call can be wrapped into it. 
This can be also used to dynamically hide or unhide available tracks in the library. 
Maybe we can use QFileSystemWatcher or a native solution using https://msdn.microsoft.com/en-us/library/windows/desktop/aa385474%28v=vs.85%29.aspx

sblaisot committed:
Add comment to explain why we don't use drive.canonicalPath()

sblaisot commented:
ok, comment added

bug filed for the remaining interface hang: https://bugs.launchpad.net/mixxx/+bug/1488046
in fact, the problem is not solely on drive.canonicalPath(), but with all attempts to actually access faulty share. After the first try, it seems the system remember the state of the drive and is much more quicker until it decides to try again. This is system dependant.

Also, the problem doesn't seems to be accessible vs not accessible, but how much time it will take to build the directory tree of any directory (either a network share or not). I can also reproduce interface hangs with a well accessible network share that has a huge number of subdirectories (it's long to enumerate and during that time, interface hangs). So the "right" fix to this new bug seems to move subdirectory enumeration to another thread, but that's beyond the scope of 1.12 I think.

kain88-de commented:
Did you find anything in the documentation of Qt why this is fixing the issue? Or did you maybe talk/write to them to see if this is an issue with out Qt-version or how windows is doing things in general?

sblaisot commented:
No, I didn't talk to Qt, but It doesn't seems to me this is a qt specific problem.
This is rather an Os behaviour: If you have an unaccessible share, each access to it is (or may be) slowed down.

Qt code is quite clear:
QDir::filePath doen't make any filesystem call because it's and absolutePath (by construction, because it is taken from QDir::drives())
http://code.qt.io/cgit/qt/qt.git/tree/src/corelib/io/qdir.cpp#n675

QDir::canonicalPath makes a filesystem call, and thus triggers the system access timeout.
http://code.qt.io/cgit/qt/qt.git/tree/src/corelib/io/qdir.cpp#n639

As long as you don't try to access an unavailable mountpoint, you don't have any problem.
When you try to access it for the first time, bad things begin to happen.
This is also true for NFS mountpoints under Linux for example (and the reason why you have hard mount and soft mount options for NFS mount depending you want complete system hang when NFS share is not available or unreadable content but no system hang).

kain88-de commented:
@sblaisot 

That is the kind of explanation that should ideally go into the commit message. It helps people in the
future to better understand why you did the changes and it also helps us reviewing the changes. It also
doesn't hurt to change the comments you made with these texts, as they actually give a complete technical reason why `QDir::canonicalPath` is a bad function to call in this case.

ywwg commented on src/library/browse/browsefeature.cpp:
```diff
@@ -68,8 +68,15 @@ BrowseFeature::BrowseFeature(QObject* parent,
+        // Using drive.filePath() instead of drive.canonicalPath() as it
+        // delay the startup too much if there is a network share mounted
+        // (drive letter assigned) but unavailable
+        QString cp = drive.filePath();
```
use a more descriptive variable name -- display_path is fine.  (I'm not sure what the difference between how the canonical path and filepath are used in the tree item.  Can you explain that?)

sblaisot committed:
improve code readability and add some comment about canonicalPath() in DEVICE_NODE

sblaisot commented:
ok, updated comment to explain a bit more why we avoid canonicalPath()

variable name changed

the treeItem contructor arguments are explained here:
https://github.com/mixxxdj/mixxx/blob/1.12/src/library/browse/browsefeature.cpp#L95

drive.canonicalPath() was used as the first argument, which is
// 1. argument represents the folder name shown in the sidebar later on

drive.filePath() is used as the second argument:
// 2. argument represents the folder path which MUST end with '/'

when expanding the item in the left-hand pane, only the folder path (argument #2) is used to access directory / files. display string is only for ... display

the old comment was wrong saying:
drive.canonicalPath(), //  displays C:

under windows seven / QT 4.8, drive.canonicalPath() displays C:/ the exact same as filePath
I introduced the new variable and chop to respect prior author willing to have C: in the device list and not C:/

ywwg commented:
LGTM, @kain88-de any more comments?

sblaisot commented:
Thanks.
